### PR TITLE
[Proof-of-concept] Basic email sending with mailgun

### PR DESCRIPTION
**Pull request by mbrandonw**
### What

Sends an email to us every time someone signs up for our pre-launch page! This can be the basis of all of our email sending! I'm sure there's more to do, but it gets the job done for now and it's pretty fun.

### CODE REVIEW

File: Sources/PointFree/EnvVars.swift [hunk 1]
```diff
@@ -32,6 +32,10 @@ enum EnvVars {
     static let clientSecret = env["GITHUB_CLIENT_SECRET"] ?? "deadbeef-client-secret"
   }
 
+  enum Mailgun {
+    static let apiKey = env["MAILGUN_PRIVATE_API_KEY"] ?? "deadbeef-mg-api-key"
```
Comment (mbrandonw) [inline Sources/PointFree/EnvVars.swift]: new private key, ive already added it to the environment vars on our host

File: Sources/PointFree/Environment.swift [hunk 2]
```diff
@@ -6,21 +6,25 @@ import Prelude
 public typealias AirtableCreateRow = (_ email: String) -> (_ baseId: String) -> EitherIO<Prelude.Unit, Prelude.Unit>
 public typealias FetchGitHubUser = (GitHubAccessToken) -> EitherIO<Prelude.Unit, GitHubUser>
 public typealias FetchAuthToken = (_ code: String) -> EitherIO<Prelude.Unit, GitHubAccessToken>
+public typealias SendEmail = (_ email: Email) -> EitherIO<Prelude.Unit, SendEmailResponse>
```
Comment (mbrandonw) [inline Sources/PointFree/Environment.swift]: new mockable endpoint for sending emails!

File: Sources/PointFree/LaunchSignup/LaunchSignup.swift [hunk 3]
```diff
@@ -25,6 +29,34 @@ private func airtableStuff<I>(_ conn: Conn<I, String>) -> IO<Conn<I, Either<Prel
   return result.map { conn.map(const($0)) }
 }
 
+func notifyUsOfNewSignup<I>(_ conn: Conn<I, String>) -> IO<Conn<I, String>> {
```
Comment (mbrandonw) [inline Sources/PointFree/LaunchSignup/LaunchSignup.swift]: middleware that takes the incoming email, sends us the notification email, and then passes the email along through the middleware.

File: Sources/PointFree/LaunchSignup/LaunchSignup.swift [hunk 4]
```diff
@@ -25,6 +29,34 @@ private func airtableStuff<I>(_ conn: Conn<I, String>) -> IO<Conn<I, Either<Prel
...
+      content: inj2(notifyUsView.view(conn.data))
+      )
+      .run
+      .perform()
+
+    return conn
+  }
+}
+
+let notifyUsView = View<String> { email in
```
Comment (mbrandonw) [inline Sources/PointFree/LaunchSignup/LaunchSignup.swift]: a view for email!

File: Sources/PointFree/LaunchSignup/LaunchSignup.swift [hunk 5]
```diff
@@ -25,6 +29,34 @@ private func airtableStuff<I>(_ conn: Conn<I, String>) -> IO<Conn<I, Either<Prel
...
+  html([
+    head([style(reset)]),
```
Comment (mbrandonw) [inline Sources/PointFree/LaunchSignup/LaunchSignup.swift]: that includes css!

File: Sources/PointFree/Mailgun/Mailgun.swift [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import Either
+import Foundation
+import HttpPipeline
+import Optics
+import Prelude
+
+enum Tracking: String {
```
Comment (mbrandonw) [inline Sources/PointFree/Mailgun/Mailgun.swift]: the contents of this file could be extracted to a lil swift mailgun package.

I lifted some mailgun parameters into proper types and made some functions for hitting the api

File: Sources/PointFree/Mailgun/Mailgun.swift [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  var trackingOpens: TrackingOpens? = nil
+  var domain: String
+}
+
+public struct SendEmailResponse: Decodable {
+  let id: String
+  let message: String
+}
+
+func mailgunSend(email: Email) -> EitherIO<Prelude.Unit, SendEmailResponse> {
```
Comment (mbrandonw) [inline Sources/PointFree/Mailgun/Mailgun.swift]: I'm not capturing the error cause we dont need it right now, but it's basically a simple wrapper around a `message: String` field...

File: Sources/PointFree/Mailgun/Mailgun.swift [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    })
+}
+
+private func attachedMailgunAuthorization(_ headers: [String: String]?) -> [String: String]? {
+  return (headers ?? [:])
+    |> key("Authorization") .~ ("Basic " + Data("api:\(EnvVars.Mailgun.apiKey)".utf8).base64EncodedString())
+}
+
+// TODO: move to swift-prelude
+private func compact<K, V>(_ xs: [K: V?]) -> [K: V] {
```
Comment (mbrandonw) [inline Sources/PointFree/Mailgun/Mailgun.swift]: @stephencelis down to add `compact` functions to prelude? I like the name!

File: Sources/PointFree/Mailgun/Mailgun.swift [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            ?? .left(unit)
+        )
+        }
+        .resume()
+    })
+}
+
+private func attachedMailgunAuthorization(_ headers: [String: String]?) -> [String: String]? {
+  return (headers ?? [:])
+    |> key("Authorization") .~ ("Basic " + Data("api:\(EnvVars.Mailgun.apiKey)".utf8).base64EncodedString())
```
Comment (mbrandonw) [inline Sources/PointFree/Mailgun/Mailgun.swift]: love the `key` lens!

File: Sources/PointFree/Mailgun/SendEmail.swift [hunk 10]
```diff
@@ -0,0 +1,56 @@
+import Either
+import Html
+import Prelude
+
+func sendEmail(
+  from: String,
+  to: [String],
+  subject: String,
+  content: Either3<String, [Node], (String, [Node])>,
```
Comment (mbrandonw) [inline Sources/PointFree/Mailgun/SendEmail.swift]: this `sendEmail` is just like a nicer interface to the Mailgun one that exposes everything their API has to offer.

a fun / interesting / weird thing we are doing here is using our new `Either3` type! it allows us to support the user sending us plain text, html (from which we will derive plain text), or both! Haskell has a `These` type for this, but honestly the more generic `Either3` seems better!

File: Sources/PointFree/Mailgun/SendEmail.swift [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  domain: String = "mg.pointfree.co"
+  )
+  -> EitherIO<Unit, SendEmailResponse> {
+
+    let (plain, html): (String, String?) =
+      destructure(
```
Comment (mbrandonw) [inline Sources/PointFree/Mailgun/SendEmail.swift]: we can provide a `destructure` on `Either3` for allowing us to handle each of the 3 cases!

File: Sources/PointFree/LaunchSignup/LaunchSignup.swift [hunk 12]
```diff
@@ -25,6 +29,34 @@ private func airtableStuff<I>(_ conn: Conn<I, String>) -> IO<Conn<I, Either<Prel
   return result.map { conn.map(const($0)) }
 }
 
+func notifyUsOfNewSignup<I>(_ conn: Conn<I, String>) -> IO<Conn<I, String>> {
+  return IO {
+
+    // Fire-and-forget to notify us that someone signed up
+    _ = sendEmail(
```
Comment (stephencelis) [inline Sources/PointFree/LaunchSignup/LaunchSignup.swift]: Let's fig out the parallel story here soon!

File: Sources/PointFree/Mailgun/SendEmail.swift [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        tracking: nil,
+        trackingClicks: nil,
+        trackingOpens: nil,
+        domain: domain
+      )
+    )
+}
+
+// TODO: move to swift-prelude
+private func destructure<A, B, C, D>(
```
Comment (mbrandonw) [inline Sources/PointFree/Mailgun/SendEmail.swift]: I have a branch on prelude for making a bunch of these, but I also went down a wild rabbit hole that I need to dig myself out of

File: Tests/PointFreeTests/TestCase.swift [hunk 14]
```diff
@@ -12,7 +12,8 @@ class TestCase: XCTestCase {
       env: .init(
         airtableStuff: const(const(pure(unit))),
         fetchAuthToken: const(pure(.init(accessToken: "deadbeef"))),
-        fetchGitHubUser: const(pure(.init(email: "hello@pointfree.co", id: 1, name: "Blob")))
+        fetchGitHubUser: const(pure(.init(email: "hello@pointfree.co", id: 1, name: "Blob"))),
+        sendEmail: const(pure(.init(id: "deadbeef", message: "success!")))
```
Comment (mbrandonw) [inline Tests/PointFreeTests/TestCase.swift]: stub out for tests!

File: Sources/PointFree/Mailgun/Mailgun.swift [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    "from": email.from,
+    "to": email.to.joined(separator: ","),
+    "cc": email.cc?.joined(separator: ","),
+    "bcc": email.bcc?.joined(separator: ","),
+    "subject": email.subject,
+    "text": email.text,
+    "html": email.html,
+    "tracking": email.tracking?.rawValue,
+    "tracking-clicks": email.trackingClicks?.rawValue,
+    "tracking-opens": email.trackingOpens?.rawValue
```
Comment (stephencelis) [inline Sources/PointFree/Mailgun/Mailgun.swift]: Should we make this stuff `Codable`?

File: Sources/PointFree/LaunchSignup/LaunchSignup.swift [hunk 16]
```diff
@@ -25,6 +29,34 @@ private func airtableStuff<I>(_ conn: Conn<I, String>) -> IO<Conn<I, Either<Prel
...
+      from: "Point-Free <brandon@pointfree.co>",
+      to: ["mbw234@gmail.com", "stephen.celis@gmail.com"],
```
Comment (mbrandonw) [inline Sources/PointFree/LaunchSignup/LaunchSignup.swift]: will change this to `support@pointfree.co` once we get our forwarding figured out

File: Sources/PointFree/LaunchSignup/LaunchSignup.swift [hunk 17]
```diff
@@ -25,6 +29,34 @@ private func airtableStuff<I>(_ conn: Conn<I, String>) -> IO<Conn<I, Either<Prel
...
+      subject: "New signup for Point-Free!",
+      content: inj2(notifyUsView.view(conn.data))
```
Comment (mbrandonw) [inline Sources/PointFree/LaunchSignup/LaunchSignup.swift]: `inj2` puts us in the `[Node]` case of `String | [Node] | (String, [Node])` lol

File: Sources/PointFree/Mailgun/Mailgun.swift (hunk 8 above)
Comment (stephencelis) [reply Sources/PointFree/Mailgun/Mailgun.swift]: Yeah! Or we could have a `mapOptional` on value!

Comment (mbrandonw) [reply Sources/PointFree/Mailgun/Mailgun.swift]: oh yeah that too. in fact, I'd be down for `mapOptional` on dictionary and adding a `compact = mapOptional(id)`

File: Sources/PointFree/Mailgun/Mailgun.swift (hunk 15 above)
Comment (mbrandonw) [reply Sources/PointFree/Mailgun/Mailgun.swift]: only complication is that the to/cc/bcc all need to be comma separated in the submitted data, or submitted in a url encoded format like `to=a@a.com&to=b@b.com&...` whereas our form encoder helper would do `to[]=a@a.com&to[]=b@b.com&...` :/ prob means we should make our encoder customizable to allow stuff like that.

File: Sources/PointFree/Mailgun/Mailgun.swift (hunk 9 above)
Comment (stephencelis) [reply Sources/PointFree/Mailgun/Mailgun.swift]: Realizing now that we can actually use key path subscripts!

``` swift
return (headers ?? [:])
  |> \.["Authorization"] .~ ("Basic " + …)
 ```

Comment (mbrandonw) [reply Sources/PointFree/Mailgun/Mailgun.swift]: amazing!

**Pull request merged**